### PR TITLE
Implement a very simple conditional templating system

### PR DESCRIPTION
**Pull request by felixarntz**
In regards to https://github.com/GoogleChromeLabs/third-party-capital/pull/71#issuecomment-2332572666: This is a _very simplistic_ (!) attempt at supporting conditionals in the templating system.

It allows using Mustache-like conditionals, e.g.
```
{{#consentValues}}window['gtag-'+{{l}}]('consent', {{consentType}}, {{consentValues}});{{/consentValues}}
```

We could use that to enhance the solution from #71, making `consentValues` null by default, so that the call is only made if this is specified. In light of not supporting more complicated conditionals, it would make sense to leave "default" as the default value for `consentType`, as the more important piece of information to provide for such calls is the configuration object anyway.

It probably has several limitations e.g. when it comes to having multiple conditional clauses of the same name in one code snippet etc., but as long as we're aware of that limitation, maybe this is an okay start for now? There are certainly more reliable and established solutions out there, but for those it would probably make more sense to pull in a library, which requires more discussion. So maybe this works? :)

(My first JS PR in this codebase, so let's see if I got things right.)

### CODE REVIEW

**Comment by felixarntz** [comment]
@huang-julien @flashdesignory Note that there's a PHPUnit test failure, which however is unrelated to the new code templating logic: For some reason, the new `consentType` and `consentValues` values are included as query parameters alongside the `https://www.googletagmanager.com/gtag/js` URL, which is clearly wrong.

Have you been able to verify that that is _not_ the case on the JS side? Potentially it's something that I implemented incorrectly in PHP, but when comparing with the JS code, it seems similar. Since there is no test coverage for the exact third-party outputs on the JS side, I just want to make sure this isn't broken on the JS side as well.

**Comment by huang-julien** [comment]
hmmm the JSON data structure should prevent that. A Script object can't have url and code at the same time.

**Comment by felixarntz** [comment]
@huang-julien Makes sense in general. But see this code: https://github.com/GoogleChromeLabs/third-party-capital/blob/main/src/utils/index.ts#L106-L113

As far as I can tell, this combines all params and optional params from all scripts that are part of the third party into one array. And further down that is passed to `formatUrl`. Maybe I misread that code somewhere, but I would think that based on that logic the JS side has the same bug at the moment.

**Comment by felixarntz** [comment]
@housseindjirdeh @flashdesignory It would be great to get your input on https://github.com/GoogleChromeLabs/third-party-capital/pull/73#issuecomment-2332709312 - this may be an already present bug that's unrelated to this PR, but is "exposed" by the PHPUnit tests?

File: src/utils/index.ts
Comment (felixarntz) [reply src/utils/index.ts:63]: @housseindjirdeh @flashdesignory Not sure whether we need to do something about this? Do you think it's relevant to address?

**Comment by huang-julien** [comment]
tbh, we don't really use formatData with nuxt script since we need to do our own transformation. 

With JS `formatData` filters all possible params into an array then applies the external's script default value if necessary. `formatUrl`  has a last `optionalParams` parameters which is where we set it if not present 
here https://github.com/GoogleChromeLabs/third-party-capital/blob/ad7e126cb1bd64b2db76cd1e2c88b7a3f9d919ac/src/utils/index.ts#L163 and https://github.com/GoogleChromeLabs/third-party-capital/blob/ad7e126cb1bd64b2db76cd1e2c88b7a3f9d919ac/src/utils/index.ts#L171

File: src/utils/index.ts
Comment (flashdesignory) [reply src/utils/index.ts:63]: hmm, sounds like it might mostly be an issue with larger input, but maybe still something we should fix. 
what was the reasoning to change the regular expression? Was it to avoid filtering out nullisch values afterwards, or did the consent object require us to change this?

**Comment by flashdesignory** [comment]
> @huang-julien Makes sense in general. But see this code: https://github.com/GoogleChromeLabs/third-party-capital/blob/main/src/utils/index.ts#L106-L113
> 
> As far as I can tell, this combines all params and optional params from all scripts that are part of the third party into one array. And further down that is passed to `formatUrl`. Maybe I misread that code somewhere, but I would think that based on that logic the JS side has the same bug at the moment.

I'll take a look on Monday. This is not really introducing a bug at the moment, but it's not the cleanest way to handle both params and optionalParams.

File: src/utils/index.ts
Comment (felixarntz) [reply src/utils/index.ts:63]: It's not a change, this is an entirely new regular expression, to account for if constructs using Mustache syntax (see PR description for more context).

I'm not sure I understand the CodeQL complaint here or how we could fix it. 🤔 FWIW it all works as expected as shown by the new tests.

**Comment by felixarntz** [comment]
> I'll take a look on Monday. This is not really introducing a bug at the moment, but it's not the cleanest way to handle both params and optionalParams.

Are you sure? From what I can tell, it is introducing a bug as of #71. The code is combining the available `params` and `optionalParams`, so far so good. But I think the problem is that it combines those params from _all_ scripts, when this should rather be happening per script. The `consentType` and `consentValues` keys are only relevant for the Analytics `code` script, not the Analytics `url` script. But since all params for all scripts are merged, this distinction is ignored, leading to code specific parameters to be included in the URL. I think that's the bug - we only haven't encountered it before since until #71 the `params` and `optionalParams` for both scripts always matched.

File: src/utils/index.ts
Comment (flashdesignory) [reply src/utils/index.ts:63]: the regex makes sense to me and seems correct (although I'm not a regex master,lol). 

Maybe we could limit the string input length? or we could throw an error earlier if it starts with '{{{{#' for some reason?

**Comment by felixarntz** [comment]
@flashdesignory Once #74 is merged, I'll iterate on this so we can hopefully get it in soon and that way unblock #71 and #72, if you're on board with this approach?

**Comment by huang-julien** [comment]
Should we add a default value for `consentType` ? We can indicate the default value with some JSDoc

**Comment by felixarntz** [comment]
> Should we add a default value for `consentType` (in this PR or in another one) ? We can indicate the default value with some JSDoc

Isn't the default values for `consentType` already set to "default"?

**Comment by huang-julien** [comment]
oops my bad, was doing some tests and forogt to add it 😅 sorry . Anyway, it's working well for me :+1:

**Review by flashdesignory — APPROVED** [review]
LGTM

**Review by adamsilverstein — APPROVED** [review]
Nice work!

**Pull request merged**
